Proofs/Experiments/AbsVCG.lean

```diff
@@ -0,0 +1,383 @@
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author(s): Shilpi Goel
+
+Experimental: Use the Correctness module to prove that this program implements
+absolute value correctly. The objective of this exercise is to determine the
+"shape" of the lemmas that are needed for automation and/or to modify the
+definitions in Correctness, if needed.
+-/
+import Arm
+import Tactics.StepThms
+import Tactics.Sym
+import Correctness.Correctness
+
+namespace AbsVCG
+
+def program : Program :=
+  def_program
+   [
+    (0x4005d0#64, 0x2a0003e1#32), --  mov w1, w0
+    (0x4005d4#64, 0x131f7c00#32), --  asr w0, w0, #31
+    (0x4005d8#64, 0x0b000021#32), --  add w1, w1, w0
+    (0x4005dc#64, 0x4a000020#32), --  eor w0, w1, w0
+    (0x4005e0#64, 0xd65f03c0#32)] --  ret
+
+/-- `Abs` structure: Field `w0` refers to the 32-bit input to the program that
+  is expected to be in register `w0` in the initial state, and `state` refers
+  to the Arm machine state.
+
+  We club the input and the state together so that we can express the
+  post-condition of the program only in terms of this structure `Abs`; see
+  `abs_post`. -/
+structure Abs where
+  w0      : BitVec 32
+  state   : ArmState
+
+/-- State Machine for the `Abs` program: the `next` function only steps the
+  `state` and leaves `w0` unmodified. -/
+instance : Sys Abs where
+  some := { w0 := 0#32, state := ArmState.default }
+  next := fun abs => { w0 := abs.w0,
+                       state := stepi abs.state }
+
+/-- Precondition for the correctness of the `Abs` program. -/
+def abs_pre (a : Abs) : Prop :=
+  read_gpr 32 0#5 a.state = a.w0 ∧
+  read_pc a.state = 0x4005d0#64 ∧
+  a.state.program = program ∧
+  read_err a.state = StateError.None
+
+
+/-- Specification of the absolute value computation for a 32-bit bitvector. -/
+def spec (x : BitVec 32) : BitVec 32 :=
+  -- We prefer the current definition as opposed to:
+  -- BitVec.ofNat 32 x.toInt.natAbs
+  -- because the above has functions like `toInt` that do not play well with
+  -- bitblasting/LeanSAT.
+  let msb := BitVec.extractLsb 31 31 x
+  if msb == 0#1 then
+    x
+  else
+    (0#32 - x)
```

Comment:
This is mildly unforunate --- can you request that we add signed compare to the BitVec API?

Reply:
Actually, BitVec does have [`slt`](https://github.com/leanprover/lean4/blob/30cf3bb3bfdac33a7ab385b25e0d262e0e3d8599/src/Init/Data/BitVec/Basic.lean#L363) and [`sle`](https://github.com/leanprover/lean4/blob/30cf3bb3bfdac33a7ab385b25e0d262e0e3d8599/src/Init/Data/BitVec/Basic.lean#L370) that correspond to SMTLIB functions `bvslt` and `bvsle` (respectively). So we could rewrite this `spec` to use `slt`.